### ISSUE TITLE
Expected continuation values can overflow to infinity when next-period probabilities span a wide range

### ISSUE THREAD

**Issue by hmgaudecker** [OpenSourceEconomics/lcm#424]
## Summary

When pylcm takes the expectation of next period's value function over a discrete
distribution of next-period states, the result can come back as `+inf` even though the
true expectation is a small, finite number. The model is well specified, the
probabilities are legitimate, and the answer is unremarkable — the computation
overflows anyway.

Once that happens the value function contains `inf`, and everything downstream of it —
policies, the backward induction into earlier periods, simulated paths — is
meaningless. It fails loudly in the sense that `inf` is visible if you look, but it
fails silently in the sense that nothing raises.

## What goes wrong

The expectation is a weighted average: each node of the next-period distribution
carries a weight, the value at that node is multiplied by its weight, the products are
summed, and the sum is divided by the total weight. The weights arrive unnormalized,
so the division by total weight is what turns them into probabilities.

The trouble is the order. The product `weight × value` is formed *before* the
normalization, so it is computed at the scale of the raw weights rather than at the
scale of the probabilities. If the weights span a wide range — one node carrying almost
all the mass, another carrying almost none — the largest product exceeds the largest
representable number and becomes `inf`. Dividing `inf` by the total weight afterwards
leaves `inf`.

## Reproduction

Four nodes with values `[4, 1, 1, 1]`. The first node is overwhelmingly the likely one,
the last is vanishingly unlikely. After normalization the distribution puts essentially
all probability on the first node, so the expectation is `4.0` exactly.

```
=== single precision ===
  weights            [8.5070592e+37 1.0000000e+00 1.0000000e+00 1.1754944e-38]
  raw weights*values [          inf 1.0000000e+00 1.0000000e+00 1.1754944e-38]
  exact mean         4.0
  production returns inf

=== double precision ===
  weights            [4.49423284e+307 1.00000000e+000 1.00000000e+000 2.22507386e-308]
  raw weights*values [            inf 1.00000000e+000 1.00000000e+000 2.22507386e-308]
  exact mean         4.0
  production returns inf
```

Two things worth noting:

- **Higher precision does not help.** It reproduces identically in single and double
  precision, because what matters is the *ratio* of the largest weight to the smallest,
  not their absolute size. Double precision simply moves the ratio at which it bites.
- **The inputs are not pathological.** These are ordinary probabilities: non-negative,
  finite, and summing to one once normalized.

## When this is likely to bite in practice

Anywhere the next-period distribution is close to degenerate — most of the mass on one
node and a long thin tail on the others:

- highly persistent discretized processes (a near-unit-root AR(1) on a fine grid, where
  a state transitions to its own neighbourhood with probability close to one);
- fine grids generally, since refining the grid pushes the off-diagonal mass down;
- small survival, entry, or exit probabilities interacted with a state transition;
- any transition matrix whose rows are near-degenerate by construction.

It affects the expectation step used in both solution and simulation, so it is not
confined to one solver.

## Suggested fix

Normalize before multiplying, rather than after. Concretely:

1. Divide the weights by the magnitude of their total mass **before** forming
   `weight × value`, so the products are formed at the scale of probabilities rather
   than at the scale of the raw weights.
2. Do that division by a power of two rather than by the mass itself. Scaling by a
   power of two is exact in floating point, so this introduces no rounding error of its
   own — it moves the exponent without touching the significant digits.
3. Bound the remaining scale adjustment by the room actually available above the
   smallest representable number, so that fixing the overflow at the top does not
   introduce an underflow at the bottom. A node that is genuinely negligible should
   contribute nothing; a node that is small but decision-relevant should still
   contribute.

The cost of this is modest. Measured on the expectation step in isolation, a candidate
implementation runs about 1.15× the current time in single precision and about 1.51× in
double. What that amounts to as a share of a full solve has not been measured yet, and
should be before anything is merged.

## A related inconsistency worth settling at the same time

A negative weight is a misspecification — a negative probability is not a probability.
At present a negative weight flows through the arithmetic and can produce a finite,
plausible-looking number, so a broken transition can be mistaken for an answer.

pylcm already guards against this one level up: the probabilities attached to regime
transitions are checked for negativity, deliberately at run time rather than at model
setup, because the weights are computed from user functions and parameters and so have
no values to inspect until the model runs. There is no equivalent guard one level down,
on the weights of the nodes *within* a distribution.

Whatever the fix does here should be consistent with that existing treatment. The
natural choice is for a negative node weight to make the result not-a-number, so that a
malformed transition cannot be laundered into a number that looks like the answer to a
well-posed question.

**Comment by hmgaudecker** [comment]
Here is a working implementation of the fix described above, as a patch. It applies
cleanly to `main` (1e81552).

It does three things:

- forms the products at the scale of probabilities rather than of the raw weights, by
  dividing out the mass's magnitude as a power of two beforehand, so no rounding error
  is introduced by the rescaling itself;
- bounds the remaining scale adjustment by the room available above the smallest
  representable number, so removing the overflow at the top does not create an
  underflow at the bottom;
- makes a negative weight poison the result, per the closing section above, and updates
  the test that previously asserted the arithmetic-passthrough behaviour.

Measured on the expectation step in isolation: 1.153x current in single precision,
1.509x in double, against a budget of 1.25x. The share of a full solve this represents
is still unmeasured and should be established before merging.

<details>
<summary>Patch</summary>

```diff
diff --git a/src/_lcm/regime_building/Q_and_F.py b/src/_lcm/regime_building/Q_and_F.py
index b87ac6cf..82ce0bb9 100644
--- a/src/_lcm/regime_building/Q_and_F.py
+++ b/src/_lcm/regime_building/Q_and_F.py
@@ -12,9 +12,12 @@ from dags import (
 
 from _lcm.certainty_equivalent import CertaintyEquivalent, LinearExpectation
 from _lcm.probability import (
+    _unbiased_exponent,
+    binades_above_smallest_normal,
     is_negative,
     is_represented_zero,
     normalized_scaled_weights,
+    scaled_by_power_of_two,
     scaled_down_by_power_of_two,
 )
 from _lcm.regime_building.next_state import (
@@ -1211,13 +1214,11 @@ def _expectation_over_stochastic_nodes(
       contribution, turning a broken transition into a plausible number.
 
     Each node arrives with a scale of its own. The denominator names the
-    weights on the largest scale all nodes reach. The numerator first forms
-    `w · v` and then applies that node's relative scale. This order is exact for
-    the scaled joint-probability coefficients supplied here: every live
-    coefficient is normal and no larger than one, so multiplying it by a
-    finite continuation cannot overflow, while applying the scale to the
-    weight first can flush a decision-relevant contribution before the value
-    supplies its binades.
+    weights on the largest scale all live nodes reach. Before forming the
+    numerator, the row mass is split into an order-one significand and an
+    integer exponent. Dividing only by that significand and folding the exponent
+    into each node's scale reserves the headroom a largest coefficient needs
+    without turning a rare coefficient into zero before its value can offset it.
 
     Args:
         values: Next period's value at this target's stochastic nodes.
@@ -1229,26 +1230,59 @@ def _expectation_over_stochastic_nodes(
 
     """
     shifts_arr = jnp.asarray(shifts)
-    common_shift = jnp.min(shifts_arr)
+
+    # Only live, finite nodes select the scale on which the row mass is read.
+    # A null event may carry an arbitrary bookkeeping shift and must not lower
+    # every live coefficient merely by existing.
+    zero_weight = is_represented_zero(weights)
+    negative_weight = is_negative(weights)
+    finite_weight = jnp.isfinite(weights)
+    invalid_weight = negative_weight | (~finite_weight & ~zero_weight)
+    live = ~zero_weight & ~negative_weight & finite_weight
+    effective_weights = jnp.where(
+        invalid_weight, jnp.asarray(jnp.nan, dtype=weights.dtype), weights
+    )
+    unusable = jnp.max(shifts_arr)
+    common_shift = jnp.min(jnp.where(live, shifts_arr, unusable))
     relative_scale = (common_shift - shifts_arr).astype(jnp.int32)
 
-    # The mass only needs the weights on one scale. A live node too far below
-    # that scale to register contributes no share to a total of order one.
-    lowered_weights = scaled_down_by_power_of_two(weights, relative_scale)
+    lowered_weights = jnp.where(
+        live,
+        scaled_down_by_power_of_two(effective_weights, relative_scale),
+        jnp.zeros_like(effective_weights),
+    )
     weight_sum = jnp.sum(lowered_weights)
     safe_weight_sum = jnp.where(weight_sum > 0.0, weight_sum, 1.0)
 
-    # The numerator is different: a tiny probability can meet a large value and
-    # make an ordinary contribution. Form the product while the coefficient is
-    # still normal, then scale the product down. Only a represented-zero
-    # coefficient is a null event; a live NaN or infinity must remain visible.
+    # Keep the mass exponent as an integer and divide only by its order-one
+    # significand.  This reserves enough numerator headroom even when all node
+    # shifts are equal, but cannot erase a rare coefficient before a large
+    # continuation value supplies the missing binades.  Reading the exponent
+    # from the scalar mass avoids P1's second array scaling pass and reduction.
+    mass_exponent = _unbiased_exponent(safe_weight_sum)
+    mass_mantissa = scaled_by_power_of_two(
+        safe_weight_sum, -mass_exponent
+    )
+    normalized_coefficients = effective_weights / mass_mantissa
+    full_scale = relative_scale - mass_exponent
+
+    room = binades_above_smallest_normal(normalized_coefficients)
+    on_weight = jnp.maximum(full_scale, -room)
+    scaled_coefficients = scaled_down_by_power_of_two(
+        normalized_coefficients, on_weight
+    )
     safe_values = jnp.where(
-        is_represented_zero(weights) & ~jnp.isfinite(values),
+        zero_weight & ~jnp.isfinite(values),
         jnp.zeros((), dtype=values.dtype),
         values,
     )
-    terms = scaled_down_by_power_of_two(weights * safe_values, relative_scale)
-    return jnp.sum(terms) / safe_weight_sum
+    products = zero_safe_weighted_term(
+        weight=scaled_coefficients,
+        value=safe_values,
+        subnormal_is_accounted_for=False,
+    )
+    terms = scaled_down_by_power_of_two(products, full_scale - on_weight)
+    return jnp.sum(terms)
 
 
 def _as_lottery(
diff --git a/tests/solution/test_scaled_probability_expectation.py b/tests/solution/test_scaled_probability_expectation.py
index a99923d3..8ac1302d 100644
--- a/tests/solution/test_scaled_probability_expectation.py
+++ b/tests/solution/test_scaled_probability_expectation.py
@@ -18,8 +18,8 @@ def _maybe_jit(
     return jax.jit(func) if compiled else func
 
 
-def test_production_reduction_avoids_general_exponent_balancing() -> None:
-    """Keep `ldexp`/`frexp` off the target-node expectation hot path."""
+def test_production_reduction_avoids_per_node_general_exponent_balancing() -> None:
+    """Keep general per-node exponent balancing off the target hot path."""
     dtype = jnp.zeros(()).dtype
     traced = jax.make_jaxpr(_expectation_over_stochastic_nodes)(
         values=jnp.ones(5, dtype=dtype),
@@ -30,7 +30,41 @@ def test_production_reduction_avoids_general_exponent_balancing() -> None:
 
     assert "name=ldexp" not in graph
     assert "name=frexp" not in graph
-    assert "clz" not in graph
+
+
+@pytest.mark.parametrize("compiled", [False, True], ids=["eager", "jit"])
+def test_row_mass_headroom_prevents_a_finite_mean_from_overflowing(
+    *, compiled: bool
+) -> None:
+    """A top-binade coefficient must not overflow before normalization."""
+    dtype = jnp.zeros(()).dtype
+    finfo = jnp.finfo(dtype)
+    top = jnp.asarray(2.0 ** (finfo.maxexp - 2), dtype=dtype)
+    tiny = jnp.asarray(finfo.tiny, dtype=dtype)
+    reduce = _maybe_jit(_expectation_over_stochastic_nodes, compiled=compiled)
+
+    got = reduce(
+        values=jnp.asarray([4.0, 1.0, 1.0, 1.0], dtype=dtype),
+        weights=jnp.asarray([top, 1.0, 1.0, tiny], dtype=dtype),
+        shifts=jnp.full(4, finfo.maxexp - 2, dtype=jnp.int32),
+    )
+
+    np.testing.assert_array_equal(np.asarray(got), np.asarray(4.0, dtype=dtype))
+
+
+@pytest.mark.parametrize("compiled", [False, True], ids=["eager", "jit"])
+def test_a_null_nodes_shift_does_not_set_the_mass_scale(*, compiled: bool) -> None:
+    """A zero-probability node carries no mass, whatever scale it stores."""
+    dtype = jnp.zeros(()).dtype
+    reduce = _maybe_jit(_expectation_over_stochastic_nodes, compiled=compiled)
+
+    got = reduce(
+        values=jnp.asarray([3.0, -jnp.inf], dtype=dtype),
+        weights=jnp.asarray([1.0, 0.0], dtype=dtype),
+        shifts=jnp.asarray([100, -100], dtype=jnp.int32),
+    )
+
+    np.testing.assert_array_equal(np.asarray(got), np.asarray(3.0, dtype=dtype))
 
 
 @pytest.mark.parametrize("compiled", [False, True], ids=["eager", "jit"])
diff --git a/tests/solution/test_zero_probability_node.py b/tests/solution/test_zero_probability_node.py
index f644f514..9eb80158 100644
--- a/tests/solution/test_zero_probability_node.py
+++ b/tests/solution/test_zero_probability_node.py
@@ -170,20 +170,15 @@ def test_the_expectation_drops_a_zero_weight_node_rather_than_multiplying_it() -
     np.testing.assert_allclose(np.asarray(got), np.asarray(1.0))
 
 
-def test_a_negative_weight_is_not_laundered_into_a_zero_contribution() -> None:
-    """A weight below zero is a malformed transition, not an impossible node.
-
-    Values `[1, 2]` at weights `[1, -1]` give `1 - 2 = -1`. Dropping the
-    negative node instead would give `1`, a value that looks like the answer to
-    a well-posed question, so the malformed weight has to carry through.
-    """
+def test_a_negative_weight_poisons_the_lottery() -> None:
+    """A negative coefficient is malformed probability mass, so the row is NaN."""
     got = _expectation_over_stochastic_nodes(
         values=jnp.asarray([1.0, 2.0]),
         weights=jnp.asarray([1.0, -1.0]),
         shifts=jnp.zeros(2, dtype=jnp.int32),
     )
 
-    np.testing.assert_allclose(np.asarray(got), np.asarray(-1.0))
+    assert bool(jnp.isnan(got))
 
 
 def test_a_nan_weight_stays_poison() -> None:
```

</details>

**Comment by hmgaudecker** [comment]
Negative weights are likely a case for poison if `log_level != "debug"` and fail with a precise message in debug mode, once more.

**Comment by hmgaudecker** [comment]
Fixed on `main` by the #390 squash commit `88ce51dc` before #400 merges.

The production certainty-equivalent reduction now represents lottery weights as coefficients plus explicit base-two shifts and splits the scale between the weight and weighted product. This avoids both failure modes described here: a top-binade intermediate cannot overflow before normalization, while a rare but decision-relevant contribution is not forced through an unrepresentable plain probability first. Invalid coefficients, including negative weights, poison their own lottery row instead of producing a plausible finite result.

The class-level regression is `tests/test_scaled_lottery_reduction_matrix.py::test_every_admissible_scaled_lottery_reduces_to_its_exact_mean`, which compares both the linear and identity quasi-arithmetic routes with exact rational arithmetic over fp32/fp64 exponent spreads, including values near the format maximum. `test_a_malformed_coefficient_poisons_its_own_row_only` covers negative and NaN weights.

#400 does not change this implementation or these regressions relative to `main`; it only consumes the established continuation machinery. Closing as completed.
